db/hardware_test.go

```diff
@@ -0,0 +1,240 @@
+func TestGetByIP(t *testing.T) {
+	TestInsertIntoDB(t)
+	var response string
+	var err error
+	a := newTestParams()
+	defer a.tinkDb.instance.Close()
+
+	if response, err = a.tinkDb.GetByIP(a.ctx.context, "192.168.1.5"); err != nil {
+		t.Fail()
+	}
+	var expected hardwarePayload
+	if err = json.Unmarshal([]byte(a.hardwareData), &expected); err != nil {
+		t.Fail()
+	}
+	var received hardwarePayload
+	if err = json.Unmarshal([]byte(response), &received); err != nil {
+		t.Fail()
+	}
+	if fmt.Sprintf("%+v", received) != fmt.Sprintf("%+v", expected) {
+		t.Logf("\nRECIEVED: %+v\nEXPECTED: %+v", received, expected)
+		t.Fail()
+	}
+}
+
+// TestGetByInstanceIP : retrieves a single hardware instances based on the IP
+// provided.
+func TestGetByInstanceIP(t *testing.T) {
+	TestInsertIntoDB(t)
+	var response string
+	var err error
+	a := newTestParams()
+	defer a.tinkDb.instance.Close()
+
+	if response, err = a.tinkDb.GetByIP(a.ctx.context, "192.168.0.1"); err != nil {
+		t.Fail()
+	}
+	var expected hardwarePayload
+	if err = json.Unmarshal([]byte(a.instanceData), &expected); err != nil {
+		t.Fail()
+	}
+	var received hardwarePayload
+	if err = json.Unmarshal([]byte(response), &received); err != nil {
+		t.Fail()
+	}
+	if fmt.Sprintf("%+v", received) != fmt.Sprintf("%+v", expected) {
+		t.Logf("\nRECIEVED: %+v\nEXPECTED: %+v", received, expected)
+		t.Logf("%s", response)
+		t.Fail()
+	}
+}
+
+// TestConcurrentInsertIntoDb : inserts five records concurrently and validates
+// whether or not all five records are actually committed.
+func TestConcurrentInsertIntoDb(t *testing.T) {
+	var err error
+	a := newTestParams()
+	defer a.tinkDb.instance.Close()
+
+	macs := []string{
+		"00-00-00-00-00-00",
+		"11-00-00-00-00-00",
+		"22-00-00-00-00-00",
+		"33-00-00-00-00-00",
+		"44-00-00-00-00-00",
+	}
+	if _, err := a.tinkDb.instance.Query("delete from hardware;"); err != nil {
+		t.Fail()
+	}
+	var wg sync.WaitGroup
+	for _, v := range macs {
+		wg.Add(1)
+		var payload hardwarePayload
+		if err = json.Unmarshal([]byte(a.hardwareData), &payload); err != nil {
+			t.Fail()
+		}
+		payload.Network.Interfaces[0].Dhcp.Mac = v
+		payload.ID = makeUUID()
+		b, err := json.Marshal(payload)
+		t.Logf("PUSHING %s", b)
+		if err != nil {
+			t.Fail()
+		}
+		go func(data string, a *testArgs, wg *sync.WaitGroup) {
+			defer wg.Done()
+			if err := a.tinkDb.InsertIntoDB(a.ctx.context, data); err != nil {
+				a.logger.Error(err)
+				t.Fail()
+			}
+		}(string(b), a, &wg)
+	}
+	wg.Wait()
+	var resultRole int
+	if err = a.tinkDb.instance.QueryRow("select count(*) as count from hardware;").Scan(&resultRole); err != nil {
+		t.Fail()
+	}
+	if resultRole != len(macs) {
+		t.Error(fmt.Errorf("MACS PUSHED: %v. \n MACS RETURNED: %v", len(macs), resultRole))
+	}
+}
+
+// TestSerializedInsertIntoDb : inserts five records one at a time and validates
+// if all five records are actually committed.
+func TestSerializedInsertIntoDb(t *testing.T) {
```

Comment:
pretty much the same as the concurrent test. Also maybe just one test with a `bool concurrent` flag to trigger go routine or not maybe?

Reply:
I wanted to keep them separate since they are testing two different scenarios. The idea is that the go tests just run by invoking `go test -v` within the db package. If I make it the same function with a boolean, then it would have to be invoked independently. Technically, I could add it to the same concurrent operation, but I felt that it made more sense keeping them two functions.

Reply:
Ah no, so what I mean is something like:
```go
func TestInsertsIntoDB(t *testing.T) {
	tests := map[string]struct{
		...
		} {	
		"concurrent 10": struct{
			concurrent: true,
			count: 10,
			waitOnChan: false, // maybe?
		},
		"serial 10": struct {
			concurrent: false,
			count: 10,
		}
	}
	...
	for _, test := range test {
		...
		if test.concurrent {
			...
			go insertIntoDB(...)
			...
		} else {
			insertIntoDB(...)
		}
		...
	}
}
```
This way we share a lot of the common setup code and the db checks code.

Reply:
better as a follow PR, maybe